odo component describe without context fails

/kind bug

## What versions of software are you using?

**Operating System:**
RHEL 8

**Output of `odo version`:**
odo v2.4.2 (88e89104e)

Server: https://api.sandbox.xyz.openshiftapps.com:6443
Kubernetes: v1.22.0-rc.0+ef241fd


## How did you run odo exactly?
created nodejs example component using IntelliJ plugin.
run odo describe command in commandline in home folder ( no context )
`odo component describe --project stephane-bouchet-dev --app app nodejs`



## Actual behavior
output failed with:
```
 ⚠  URLs couldn't not be retrieved: unable to list routes: the component name, the app name or both are empty
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x88 pc=0x263c7a5]

goroutine 1 [running]:
github.com/openshift/odo/pkg/envinfo.(*EnvInfo).ListStorage(0xc000931d70, 0x0, 0x0, 0x0, 0x0, 0x0)
        /builddir/build/BUILD/gocode/src/github.com/openshift/odo/pkg/envinfo/storage.go:149 +0xa5
github.com/openshift/odo/pkg/component.(*ComponentFullDescription).loadStoragesFromClientAndLocalConfig(0xc000358b00, 0xc00086e510, 0x3603638, 0xc000931d40, 0x7fff5198e78d, 0x6, 0x7fff5198e789, 0x3, 0xc0003d06c0, 0x0, ...)
        /builddir/build/BUILD/gocode/src/github.com/openshift/odo/pkg/component/component_full_description.go:73 +0x103
github.com/openshift/odo/pkg/component.NewComponentFullDescriptionFromClientAndLocalConfigProvider(0xc00086e510, 0xc000931d40, 0x7fff5198e78d, 0x6, 0x7fff5198e789, 0x3, 0x7fff5198e76e, 0x14, 0xc00005c0c4, 0x38, ...)
        /builddir/build/BUILD/gocode/src/github.com/openshift/odo/pkg/component/component_full_description.go:199 +0x905
github.com/openshift/odo/pkg/odo/cli/component.(*DescribeOptions).Run(0xc00000e1b0, 0xc0000d8780, 0x0, 0x0)
        /builddir/build/BUILD/gocode/src/github.com/openshift/odo/pkg/odo/cli/component/describe.go:68 +0x9b
github.com/openshift/odo/pkg/odo/genericclioptions.GenericRun(0x35c0410, 0xc00000e1b0, 0xc0000d8780, 0xc0000bcff0, 0x1, 0x5)
        /builddir/build/BUILD/gocode/src/github.com/openshift/odo/pkg/odo/genericclioptions/runnable.go:97 +0x479
github.com/openshift/odo/pkg/odo/cli/component.NewCmdDescribe.func1(0xc0000d8780, 0xc0000bcff0, 0x1, 0x5)
        /builddir/build/BUILD/gocode/src/github.com/openshift/odo/pkg/odo/cli/component/describe.go:96 +0x5e
github.com/spf13/cobra.(*Command).execute(0xc0000d8780, 0xc0000bcfa0, 0x5, 0x5, 0xc0000d8780, 0xc0000bcfa0)
        /builddir/build/BUILD/gocode/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:856 +0x2c2
github.com/spf13/cobra.(*Command).ExecuteC(0xc000985680, 0xc00005e110, 0x2ec1ec0, 0x49a2488)
        /builddir/build/BUILD/gocode/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:960 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
        /builddir/build/BUILD/gocode/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:897
github.com/spf13/cobra.(*Command).ExecuteContext(...)
        /builddir/build/BUILD/gocode/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:890
main.main()
        /builddir/build/BUILD/gocode/src/github.com/openshift/odo/cmd/odo/odo.go:70 +0x426
```

## Expected behavior
should be possible to have the correct description like
```
Component Name: nodejs
Type: nodejs
Environment Variables:
 · PROJECTS_ROOT=/project
 · PROJECT_SOURCE=/project
 · DEBUG_PORT=5858
Storage:
 · m1 of size 1Gi mounted to /tmp
URLs:
 · http://http-3000-app-stephane-bouchet-dev.apps.sandbox.x8i5.p1.openshiftapps.com exposed via 3000

```


## Any logs, error output, etc?

--- a/pkg/component/component_full_description.go
+++ b/pkg/component/component_full_description.go
@@ -47,36 +47,6 @@ func (cfd *ComponentFullDescription) copyFromComponentDesc(component *Component)
 	return json.Unmarshal(d, cfd)
 }
 
-// loadStoragesFromClientAndLocalConfig collects information about storages both locally and from the cluster.
-func (cfd *ComponentFullDescription) loadStoragesFromClientAndLocalConfig(client kclient.ClientInterface, configProvider localConfigProvider.LocalConfigProvider, componentName string, applicationName string, componentDesc *Component) error {
-	var storages storage.StorageList
-	var err error
-
-	// if component is pushed call ListWithState which gets storages from localconfig and cluster
-	// this result is already in mc readable form
-	if componentDesc.Status.State == StateTypePushed {
-		storageClient := storage.NewClient(storage.ClientOptions{
-			Client:              client,
-			LocalConfigProvider: configProvider,
-		})
-
-		storages, err = storageClient.List()
-		if err != nil {
-			return err
-		}
-	} else {
-		// otherwise simply fetch storagelist locally
-		storageLocal, err := configProvider.ListStorage()
-		if err != nil {
-			return err
-		}
-		// convert to machine readable format
-		storages = storage.ConvertListLocalToMachine(storageLocal)
-	}
-	cfd.Spec.Storage = storages
-	return nil
-}
-
 // fillEmptyFields fills any fields that are empty in the ComponentFullDescription
 func (cfd *ComponentFullDescription) fillEmptyFields(componentDesc Component, componentName string, applicationName string, projectName string) {
 	// fix missing names in case it is in not in description

--- a/pkg/component/component_full_description.go
+++ b/pkg/component/component_full_description.go
@@ -3,6 +3,7 @@ package component
 import (
 	"encoding/json"
 	"fmt"
+	v1 "k8s.io/api/apps/v1"
 	"path/filepath"
 
 	"github.com/redhat-developer/odo/pkg/kclient"
@@ -78,7 +79,7 @@ func (cfd *ComponentFullDescription) loadStoragesFromClientAndLocalConfig(client
 
 // fillEmptyFields fills any fields that are empty in the ComponentFullDescription
 func (cfd *ComponentFullDescription) fillEmptyFields(componentDesc Component, componentName string, applicationName string, projectName string) {
-	//fix missing names in case it in not in description
+	// fix missing names in case it in not in description
 	if len(cfd.Name) <= 0 {
 		cfd.Name = componentName
 	}
@@ -128,6 +129,7 @@ func NewComponentFullDescriptionFromClientAndLocalConfigProvider(client kclient.
 		return cfd, err
 	}
 	cfd.Status.State = state
+	var deployment *v1.Deployment
 	if state == StateTypePushed {
 		componentDescFromCluster, e := getRemoteComponentMetadata(client, componentName, applicationName, false, false)
 		if e != nil {
@@ -136,6 +138,12 @@ func NewComponentFullDescriptionFromClientAndLocalConfigProvider(client kclient.
 		cfd.Spec.Env = componentDescFromCluster.Spec.Env
 		cfd.Spec.Type = componentDescFromCluster.Spec.Type
 		cfd.Status.LinkedServices = componentDescFromCluster.Status.LinkedServices
+		// Obtain the deployment to correctly instantiate the Storage and URL client and get information
+		deployment, err = client.GetOneDeployment(componentName, applicationName)
+		if err != nil {
+			// ideally the error should not occur since we have already established that the component exists on the cluster
+			return cfd, err
+		}
 	}
 
 	for _, link := range configLinks {
@@ -177,6 +185,7 @@ func NewComponentFullDescriptionFromClientAndLocalConfigProvider(client kclient.
 		LocalConfigProvider: configProvider,
 		Client:              client,
 		IsRouteSupported:    routeSupported,
+		Deployment:          deployment,
 	})
 
 	urls, err = urlClient.List()
@@ -185,10 +194,21 @@ func NewComponentFullDescriptionFromClientAndLocalConfigProvider(client kclient.
 	}
 	cfd.Spec.URL = urls
 
-	err = cfd.loadStoragesFromClientAndLocalConfig(client, configProvider, componentName, applicationName, &componentDesc)
+	// Obtain Storage information
+	var storages storage.StorageList
+	storageClient := storage.NewClient(storage.ClientOptions{
+		Client:              client,
+		LocalConfigProvider: envInfo,
+		Deployment:          deployment,
+	})
+
+	// We explicitly do not fill in Spec.Storage for describe because,
+	// it is essentially a list of Storage names, which seems redundant when the detailed StorageSpec is available
+	storages, err = storageClient.List()
 	if err != nil {
-		return cfd, err
+		log.Warningf("Storages couldn't not be retrieved: %v", err)
 	}
+	cfd.Spec.Storage = storages
 
 	return cfd, nil
 }

--- a/pkg/component/component_full_description.go
+++ b/pkg/component/component_full_description.go
@@ -79,7 +79,7 @@ func (cfd *ComponentFullDescription) loadStoragesFromClientAndLocalConfig(client
 
 // fillEmptyFields fills any fields that are empty in the ComponentFullDescription
 func (cfd *ComponentFullDescription) fillEmptyFields(componentDesc Component, componentName string, applicationName string, projectName string) {
-	// fix missing names in case it in not in description
+	// fix missing names in case it is in not in description
 	if len(cfd.Name) <= 0 {
 		cfd.Name = componentName
 	}

--- a/pkg/envinfo/storage.go
+++ b/pkg/envinfo/storage.go
@@ -56,7 +56,7 @@ func (ei *EnvInfo) ValidateStorage(storage localConfigProvider.LocalStorage) err
 		return nil
 	}
 
-	//check if specified container exists or not (if specified)
+	// check if specified container exists or not (if specified)
 	containerExists := false
 	containers, err := ei.GetContainers()
 	if err != nil {
@@ -91,7 +91,7 @@ func (ei *EnvInfo) GetStorage(name string) (*localConfigProvider.LocalStorage, e
 
 // CreateStorage sets the storage related information in the local configuration
 func (ei *EnvInfo) CreateStorage(storage localConfigProvider.LocalStorage) error {
-	//initialize volume mount and volume container
+	// initialize volume mount and volume container
 	vm := []devfilev1.VolumeMount{
 		{
 			Name: storage.Name,

--- a/pkg/storage/kubernetes.go
+++ b/pkg/storage/kubernetes.go
@@ -187,7 +187,7 @@ func (k kubernetesClient) ListFromCluster() (StorageList, error) {
 
 // List lists pvc based Storage and local Storage with respective states
 func (k kubernetesClient) List() (StorageList, error) {
-	if k.localConfigProvider == nil {
+	if !k.localConfigProvider.Exists() {
 		return StorageList{}, fmt.Errorf("no local config was provided")
 	}
 

--- a/pkg/storage/kubernetes_test.go
+++ b/pkg/storage/kubernetes_test.go
@@ -619,7 +619,7 @@ func Test_kubernetesClient_List(t *testing.T) {
 			mockLocalConfig.EXPECT().GetName().Return(tt.fields.generic.componentName).AnyTimes()
 			mockLocalConfig.EXPECT().GetApplication().Return(tt.fields.generic.appName).AnyTimes()
 			mockLocalConfig.EXPECT().ListStorage().Return(tt.returnedLocalStorage, nil)
-
+			mockLocalConfig.EXPECT().Exists().Return(true)
 			tt.fields.generic.localConfigProvider = mockLocalConfig
 
 			k := kubernetesClient{

--- a/pkg/url/kubernetes.go
+++ b/pkg/url/kubernetes.go
@@ -87,7 +87,7 @@ func (k kubernetesClient) List() (URLList, error) {
 	}
 
 	localMap := make(map[string]URL)
-	if k.localConfigProvider != nil {
+	if k.localConfigProvider.Exists() {
 		// get the URLs present on the localConfigProvider
 		localURLS, err := k.localConfigProvider.ListURLs()
 		if err != nil {

--- a/pkg/url/kubernetes_test.go
+++ b/pkg/url/kubernetes_test.go
@@ -404,7 +404,7 @@ func Test_kubernetesClient_List(t *testing.T) {
 
 			mockLocalConfig := localConfigProvider.NewMockLocalConfigProvider(ctrl)
 			mockLocalConfig.EXPECT().ListURLs().Return(tt.returnedLocalURLs, nil)
-
+			mockLocalConfig.EXPECT().Exists().Return(true)
 			fkclient, fkclientset := kclient.FakeNewWithIngressSupports(true, false)
 			fkclient.Namespace = "default"
 

--- a/pkg/url/status_test.go
+++ b/pkg/url/status_test.go
@@ -161,6 +161,7 @@ func TestGetURLsForKubernetes(t *testing.T) {
 			mockLocalConfig := localConfigProvider.NewMockLocalConfigProvider(ctrl)
 			mockLocalConfig.EXPECT().GetName().Return(componentName).AnyTimes()
 			mockLocalConfig.EXPECT().GetApplication().Return("app")
+			mockLocalConfig.EXPECT().Exists().Return(true)
 			mockLocalConfig.EXPECT().ListURLs().Return(tt.envURLs, nil)
 
 			// Initialising the fakeclient

--- a/tests/integration/devfile/cmd_devfile_describe_test.go
+++ b/tests/integration/devfile/cmd_devfile_describe_test.go
@@ -32,12 +32,13 @@ var _ = Describe("odo devfile describe command tests", func() {
 		var (
 			compName = "cmp-git"
 			compType = "django"
+			appName  = "testing"
 		)
 		BeforeEach(func() {
 			// Using Django example here because it helps to distinguish between language and projectType.
 			// With nodejs, both projectType and language is nodejs, but with python-django, django is the projectType and python is the language
 			helper.CopyExample(filepath.Join("source", "python"), commonVar.Context)
-			helper.Cmd("odo", "create", "python-django", compName, "--project", commonVar.Project, "--context", commonVar.Context, "--app", "testing").ShouldPass()
+			helper.Cmd("odo", "create", "python-django", compName, "--project", commonVar.Project, "--context", commonVar.Context, "--app", appName).ShouldPass()
 			helper.Cmd("odo", "url", "create", "url-1", "--port", "3000", "--host", "example.com", "--context", commonVar.Context).ShouldPass()
 			helper.Cmd("odo", "url", "create", "url-2", "--port", "4000", "--host", "example.com", "--context", commonVar.Context).ShouldPass()
 			helper.Cmd("odo", "storage", "create", "storage-1", "--size", "1Gi", "--path", "/data1", "--context", commonVar.Context).ShouldPass()
@@ -96,6 +97,17 @@ var _ = Describe("odo devfile describe command tests", func() {
 			})
 			It("should describe the component correctly", func() {
 				checkDescribe("Pushed")
+			})
+			It("should describe the component correctly when run outside the context directory", func() {
+				By("getting human readable output", func() {
+					output := helper.Cmd("odo", "describe", compName, "--app", appName, "--project", commonVar.Project).ShouldPass().Out()
+					helper.MatchAllInOutput(output, []string{compName, compType})
+				})
+				By("getting json output", func() {
+					output := helper.Cmd("odo", "describe", compName, "--app", appName, "--project", commonVar.Project, "-o", "json").ShouldPass().Out()
+					values := gjson.GetMany(output, "kind", "metadata.name", "spec.type", "status.state")
+					Expect(helper.GjsonMatcher(values, []string{"Component", compName, compType, "Pushed"})).To(Equal(true))
+				})
 			})
 		})
 	})

